web/src/glossary.tsx

```diff
@@ -0,0 +1,54 @@
+import { createContext, useContext } from "react";
+
+type GlossaryItem = {
+    term: string;
+    pageid: string;
+    contents: string;
+};
+
+export type Glossary = Map<string, GlossaryItem>;
+
+export const GlossaryContext = createContext<{g: Glossary, r: RegExp} | null>(null);
+
+// A component which wraps arbitrary html in a span, and injects glossary terms
+// into it as hoverable pop-up links. The text is immediately rendered normally,
+// but after the glossary is loaded (which happens once per page, asynchronously),
+// the glossary terms are replaced with elements.
+
+export const GlossarySpan: React.FC<{content: string}> = ({content}) => {
+
+    const g = useContext(GlossaryContext);
+
+    // If the glossary hasn't loaded yet, just render the text normally.
+    if (g == null) {
+        return <span dangerouslySetInnerHTML={{__html: content}} />;
+    }
+
+    const glossary = g.g;
+    const glossaryRegex = g.r;
+
+    // Otherwise, replace glossary terms with links. We can do this in
+    // O(n * sum of term lengths) by finding String.prototype.indexOf of
+    // each term in the glossary (since that'd probably be backed by KMP)
+    // but I think it should be faster to compile a regex state machine
+    // once and use that instead.
```

Comment:
what about substrings? i.e. say you have 'good', 'goodness', 'goodhart's law', 'not very good' and 'blagoodblabla' as glossary entries. How would it handle the following phrases:

'this is good'
'the greater good is always the best'
'make sure goodharts law doesn't bite you'
'it's not very good if the word good get's inserted twice here'
'goodness me! Which definition of good should have been used here?'
'ritegoody is not how you describe how well someone can spell'

It might suffice to just sort the phrases by length before creating them, but you might want to check.

Reply:
Tested: word breaks and pre-sorting do work.